libplatsupport/src/mach/imx/gpt.c

```diff
@@ -217,9 +217,22 @@ static void gpt_handle_irq(void *data, ps_irq_acknowledge_fn_t acknowledge_fn, v
 
 uint64_t gpt_get_time(gpt_t *gpt)
 {
+    // Rollover of 32-bit counter can happen while we are reading it,
+    // We need to read in the following order (volatile modifier on gpt_map should ensure this):
+    // rollover_status bit
+    // gptcnt
+    // rollover_status bit.
+    // If rollover_status bit has changed, then the rollover happened sometime between the first
+    // read and the third read.  In this case conclude that the low_bits are 0.
+    uint32_t rollover_st = gpt->gpt_map->gptsr & BIT(ROV);
     uint32_t low_bits = gpt->gpt_map->gptcnt;
+    uint32_t rollover_st2 = gpt->gpt_map->gptsr & BIT(ROV);
+    if (rollover_st != rollover_st2) {
+        low_bits = 0;
+    }
+
     uint32_t high_bits = gpt->high_bits;
-    if (gpt->gpt_map->gptsr) {
+    if (rollover_st2) {
```

Comment:
Why do we do this check and then increment the value, but why? Can't we just take `high_bits` as it is now? If there had been an overflow, then `low_bits` is 0 here, which seem perfectly fine.

Reply:
For this particular timer, the counter register is 32bits and the `gpt->high_bits` is a memory region managed by the driver that counts the number of overflows.  When the 32bit counter overflows the ROV bit in the status register is set and an interrupt is generated.  The timer interrupt handler then increments high_bits and clears the status bit.  What's happening here is that the current function effectively snoops the ROV status bit to detect if a rollover has happened but the high_bits hasn't been incremented yet. If so then it increments a local copy of high_bits but doesn't reset the ROV status bit leaving it to the IRQ routine that would run immediately after.